crimsobot/cogs/cringo.py

```diff
@@ -1,225 +1,423 @@
+            self.used_emoji.extend(emojis_this_turn[0])
+            self.multiplier = self.total_turns + 1 - self.turn
+            await self.send_emojis(emojis_this_turn)
+
+            # set up listener for players scoring their cards
+            message_handler = CringoMessageHandler(ctx, timeout=self.turn_timer)
+            message_handler.set_arguments(game=self)
+
             await ctx.gather_events('on_message', handler=message_handler)
 
-            # end of turn, time to score matches
-            for player in list_of_players:
-                await cringo.cringo_score(player, turn, multiplier)
+            # end of turn, time to score everything
+            for player in self.players:
+                await cringo.cringo_score(player, self.turn, self.multiplier)  # type: ignore
 
-            turn += 1
-            if turn > total_turns:
-                embed = c.crimbed(
-                    title=None,
-                    descr=f'Game over! Check the final score in {ctx.channel.mention}!',
-                    color_name=CRINGO_RULES['color'][card_size],
-                )
-            else:
-                embed = c.crimbed(
-                    title=None,
-                    descr=f"Time's up! Round {turn} incoming.\nCheck the score in {ctx.channel.mention}!",
-                    color_name=CRINGO_RULES['color'][card_size],
-                )
+            self.turn += 1
 
-            # remove players with excessive mismatches
-            for player in list_of_players:
-                if player.mismatch_count < 8:
-                    try:
-                        await player.user.send(embed=embed)
-                    except discord.errors.Forbidden:
-                        await cringo.player_remove(list_of_players, player)
-                else:
-                    embed = await cringo.player_remove(list_of_players, player)
-                    await ctx.send(embed=embed)
+            # remove players who have too many mismatches
+            turn_embed = self.generate_end_of_turn_embed()
+            for player in self.players.copy():
+                if player.mismatch_count >= 8:
+                    left_game_embed = self.remove_from_game(player.user.id)
+                    await ctx.send(embed=left_game_embed)
+                    continue
+
+                # tell players what happens at the end of this turn
+                try:
+                    await player.user.send(embed=turn_embed)
+                except discord.errors.Forbidden:
+                    self.remove_from_game(player.user.id)
 
         # final score + awards time!
         # nerf calculated such that division by zero never attained within player limit
-        x = len(list_of_players)
+        x = len(self.players)
         nerf = 0.05*x**2 - 2.05*x + 52  # (points / nerf = coin)
 
         # check if crimsoBOT home server
         if ctx.guild and ctx.guild.id == 552650672965943296:
```

Comment:
ugh we should move this magic number somewhere

Reply:
since it's calculated by final player count it may not be worth moving it somewhere else - but perhaps a generation function of some sort could be warranted?

Reply:
no the guild id

Reply:
config sound good?

Reply:
oh it was just a mental note, doesn't need to be in this PR (and IMO shouldn't be, there are probably a few usages that we should pull out - would be good in another changeset)